src/scheduler/pbs_bitmap.c

```diff
@@ -0,0 +1,334 @@
+		return NULL;
+	}
+	
+	if(bm->bits != NULL) {
+		int i;
+		for(i = 0; i < prev_longs; i++)
+			tmp_bits[i] = bm->bits[i];
+		
+		free(bm->bits);
+	}
+	bm->bits = tmp_bits;
+		
+	return bm;
+}
+
+/* pbs_bitmap destructor */
+void
+pbs_bitmap_free(pbs_bitmap *bm)
+{
+	if(bm == NULL)
+		return;
+	free(bm->bits);
+	free(bm);
+}
+
+/**
+ * @brief turn a bit on for a bitmap
+ * @param pbm - the bitmap
+ * @param bit - which bit to turn on
+ * @return nothing
+ */
+int
+pbs_bitmap_bit_on(pbs_bitmap *pbm, long bit)
+{
+	long long_ind;
+	unsigned long b;
+	
+	if(pbm == NULL)
+		return 0;
+	
+	if (bit >= pbm->num_bits) {
+		if(pbs_bitmap_alloc(pbm, bit + 1) == NULL)
+			return 0;
+	}
+	
+	long_ind = bit / BYTES_TO_BITS(sizeof(unsigned long));
+	b = 1UL << (bit % BYTES_TO_BITS(sizeof(unsigned long)));
+	
+	pbm->bits[long_ind] |= b;
+	return 1;
+}
+
+/**
+ * @brief turn a bit off for a bitmap
+ * @param pbm - the bitmap
+ * @param bit - the bit to turn off
+ * @return nothing
+ */
+int
+pbs_bitmap_bit_off(pbs_bitmap *pbm, long bit)
+{
+	long long_ind;
+	unsigned long b;
+	
+	if (pbm == NULL)
+		return 0;
+	
+	if (bit >= pbm->num_bits) {
+		if(pbs_bitmap_alloc(pbm, bit + 1) == NULL)
+			return 0;
+	}
+	
+	long_ind = bit / BYTES_TO_BITS(sizeof(unsigned long));
+	b = 1UL << (bit % BYTES_TO_BITS(sizeof(unsigned long)));
+	
+	pbm->bits[long_ind] &= ~b;
+	return 1;
+}
+
+/**
+ * @brief get the value of a bit
+ * @param pbm - the bitmap
+ * @param bit - which bit to get the value of
+ * @return int
+ * @retval 1 if the bit is on
+ * @retval 0 if the bit is off
+ */
+int
+pbs_bitmap_get_bit(pbs_bitmap *pbm, unsigned long bit)
+{
+	long long_ind;
+	unsigned long b;
+	
+	if (pbm == NULL)
+		return 0;
+	
+	if (bit >= pbm->num_bits)
+		return 0;
+	
+	long_ind = bit / BYTES_TO_BITS(sizeof(unsigned long));
+	b = 1UL << (bit % BYTES_TO_BITS(sizeof(unsigned long)));
+	
+	return (pbm->bits[long_ind] & b) ? 1 : 0;
+}
+
+/**
+ * @brief starting at a bit, get the next on bit
+ * @param pbm - the bitmap
+ * @param start_bit - which bit to start from
+ * @return int
+ * @retval number of next on bit
+ * @retval -1 if there isn't a next on bit
+ */
+int
+pbs_bitmap_next_on_bit(pbs_bitmap *pbm, long start_bit)
+{
+	long long_ind;
+	long bit;
+	int i;
+	
+	if (pbm == NULL)
+		return -1;
+	
+	if (start_bit >= pbm->num_bits)
+		return -1;
+	
+	long_ind = start_bit / BYTES_TO_BITS(sizeof(unsigned long));
+	bit = start_bit % BYTES_TO_BITS(sizeof(unsigned long));
+
+	/* special case - look at first long that contains start_bit */
+	if (pbm->bits[long_ind] != 0) {
+		for (i = bit + 1; i < BYTES_TO_BITS(sizeof(unsigned long)); i++) {
+			if (pbm->bits[long_ind] & (1UL << i)) {
+				return (long_ind * BYTES_TO_BITS(sizeof(unsigned long)) + i);
+			}
+		}
+
+		/* didn't find an on bit after start_bit_index in the long that contained start_bit */
+		if (long_ind < pbm->num_longs) {
+			long_ind++;
+		}
+	}
+
+	for( ; long_ind < pbm->num_longs && pbm->bits[long_ind] == 0; long_ind++)
+		;
```

Comment:
I think using a while loop
```
while (long_ind < pbm->num_longs && pbm->bits[long_ind] == 0)
    long_ind++;
```
is better here.

Reply:
I disagree.  We're iterating over an array.  for() loops should be used for iteration.